### PR TITLE
fix(effects): register functional effects from object without prototype

### LINKED ISSUE #3972: "Cannot read properties of undefined (reading 'name') at isClassInstance" when importing effects from a library
### Which @ngrx/* package(s) are the source of the bug?

effects

### Minimal reproduction of the bug/regression with instructions

1. Import into your app a library that exports a provider with effects;
2. Run your app.
3. ERROR TypeError: Cannot read properties of undefined (reading 'name')
    at isClassInstance (ngrx-effects.mjs:135:29)

### Minimal reproduction of the bug/regression with instructions

```diff
// node_modules/@ngrx/effects/fesm2022/ngrx-effects.mjs
function isClassInstance(obj) {
-     return (obj.constructor.name !== 'Object' && obj.constructor.name !== 'Function');
+     return (obj.constructor && obj.constructor.name !== 'Object' && obj.constructor.name !== 'Function');
}

function mergeEffects(sourceInstance, globalErrorHandler, effectsErrorHandler) {
-    const sourceName = getSourceForInstance(sourceInstance).constructor.name;
+    const sourceName = getSourceForInstance(sourceInstance)
+      ? getSourceForInstance(sourceInstance).constructor.name
+      : null;
```

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

NgRx: ^16.1.0
Angular: ^16.1.7
Node: 18.17.0

### Other information

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #3972

## What is the new behavior?

It's possible to import functional effects that grouped via named imports from buildable libraries.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/effects/spec/effect_sources.spec.ts [hunk 1]
```diff
@@ -207,6 +208,10 @@ describe('EffectSources', () => {
       const recordB = {
         b: createEffect(() => alwaysOf(b), { functional: true }),
       };
+      const recordC = Object.freeze({
```
Comment (timdeschryver) [inline modules/effects/spec/effect_sources.spec.ts:211]: I think it's useful to add a comment here on why this record/test exists.

File: modules/effects/src/effects_resolver.ts [hunk 2]
```diff
@@ -13,7 +13,8 @@ export function mergeEffects(
   globalErrorHandler: ErrorHandler,
   effectsErrorHandler: EffectsErrorHandler
 ): Observable<EffectNotification> {
-  const sourceName = getSourceForInstance(sourceInstance).constructor.name;
+  const source = getSourceForInstance(sourceInstance);
+  const sourceName = source ? source.constructor.name : 'Object';
```
Comment (timdeschryver) [inline modules/effects/src/effects_resolver.ts:17]: Unrelated, but seeing this made me wonder how `sourceName` is used throughout the whole flow.
I ended up that it's only used to throw an error when an effect doesn't dispatch an action (an object with `type` as property).
Do we want to change the message for functional effects?
It currently throws the following message:

```js
ERROR Error: Effect "Object.logDispatchedActions()" dispatched an invalid action: {}
``` 

We can prettify this a little bit to

```js
//           👇 does this add value?
//                                👇 remove `Object.`
ERROR Error: (Functional) Effect "logDispatchedActions()" dispatched an invalid action: {}
```

**Review by brandonroberts — APPROVED** [review]

File: modules/effects/src/effects_resolver.ts (hunk 2 above)
Comment (markostanimirovic) [reply modules/effects/src/effects_resolver.ts:17]: Good point!

**Review by timdeschryver — APPROVED** [review]
👍

**Pull request merged**